### ISSUE TITLE
viewmodel based navigation as first class citzen for all platforms

### ISSUE THREAD

**Issue by ghuntley** [reactiveui/Sextant#65]
Our routing is something that should be overhauled in the future. It is far too prescriptive and platform-specific. ReactiveUI should really provide a general routing mechanism that cares not about how you manifest it in the view. All that logic should be written ​_once_​ (instead of per platform like we have now) and shared with simple shims to integrate at the UI level.

This is a great opportunity to look around and see how others are doing it. @rid00z has mentioned over beers he has suggestions from implementing this type of logic as part of his framework FreshMvvm. 

Additionally MvvmCross has the notion of handling this behavior via Presenters
- http://gregshackles.com/presenters-in-mvvmcross-a-primer/
- http://gregshackles.com/presenters-in-mvvmcross-navigating-android-with-fragments/

@jcmm33's team has done a significant amount of work on-top of ReactiveUI and have implemented their own presenter solution. Maybe this is a chance to use this implementation as the golden standard and off-load the burden from his team of having to maintain their own.

Suggested leads for this topic are @jcmm33, @rid00z, @codemillmatt and of course @kentcb / @flagbug

**Comment by weitzhandler** [comment]
Hi all,
As a swear by Prism dev, I find that RxUI doesn't provide enough in terms of navigation and routing. I actually liked the way it worked in Prism with DI `INavigationAware` and `INavigationService`, but `IScreen` and `IRoutingViewModel` sounds good too.
Can we abstract them out into a separate non-`Xamarin.Forms` dependent package, to allow for extension and implementation in other platforms?

**Comment by glennawatson** [comment]
Yeah that's the plan, the work is on-going to separate it out at the moment.

**Comment by worldbeater** [comment]
> I actually liked the way it worked in Prism with DI INavigationAware and INavigationService, but IScreen and IRoutingViewModel sounds good too.
Can we abstract them out into a separate non-Xamarin.Forms dependent package, to allow for extension and implementation in other platforms?

Probably missing the context, by what do you mean by that? Currently, `IScreen`, `RoutingState` and `IRoutableViewModel` interfaces are located in the ReactiveUI core package, and are compatible not only with Xamarin.Forms, but also with UWP, WPF, Avalonia, etc. (`RoutedViewHost`s implementations exist for almost every platform except Android and iOS)

**Comment by weitzhandler** [comment]
@worldbeater I see now, I didn't know that. Will make more research on that. Thanks.

**Comment by weitzhandler** [comment]
So what `Sextant` does is just providing the XF `NavigationPage` and VM registration?

**Comment by glennawatson** [comment]
It's a work in progress rewrite of the navigation stack inside RxUI, but the old navigation stack works as well until it's finished. Given it's very work in progress it's likely to change dramatically.

**Comment by weitzhandler** [comment]
I'm gonna start off with `IScreen` and `IRoutableViewModel`. As new things will come in I'll move on.
BTW, I'm using Splat. Can't afford to give up on DI sorry.

**Comment by glennawatson** [comment]
We added the DI containers to allow people to have a choice :) -- so both the traditional and sextant approach will interop with the registered splat DI container

**Comment by weitzhandler** [comment]
Awesome then!
